Lib/asyncore.py

```diff
@@ -399,6 +403,9 @@ def close(self):
                 if why.args[0] not in (ENOTCONN, EBADF):
                     raise
 
+    def fileno(self):
+        return self.socket.fileno()
```

Comment:
This is compatible with 2.7 and older 3.x, when this class had __getattr__, and fileno() was delegated to the self.socket.

But I found that we have incompatible implementation in file_wrapper:

```python
        def close(self):
            if self.fd < 0:
                return
            os.close(self.fd)
            self.fd = -1

        def fileno(self):
            return self.fd
```

I think we should have consistent implementation of fileno(), I would check how it behave in regular file objects and make both implementations the same. Both should return -1 when you close the fd, or raise EBADF. For backward compatibility raising EBADF should be less risky.

Finally, using self.socket means it will raise AttributeError if the socket is None - this was fixed lately in python 3 in close(), by checking if the socket is None. If we go with socket.fileno(), we need to check for None and raise EBADF.